### PR TITLE
convert.py xgen support

### PR DESCRIPTION
**Pull request by tmm1**
working toward converting XGen-7B to ggml

### CODE REVIEW

**Comment by tmm1** [comment]
currently erroring out here:

```
  File "llama.cpp/convert.py", line 986, in check_vocab_size
    raise Exception(msg)
Exception: Vocab size mismatch (model has 51200, but models/xgen-7b-8k-base has 50313).
```

if I neuter `check_vocab_size` it will continue and seems to create a ggml bin with the vocabulary included.

**Comment by clyang** [comment]
I've tried to use this PR to convert xgen-4k to GGML. It worked during the conversion but failed on loading it to inference. Here is the error message:
```
main: build = 796 (31cfbb1)
main: seed  = 1688621194
llama.cpp: loading model from models/xgen-4k-7b/ggml-model-f16.bin
error loading model: unexpectedly reached end of file
llama_load_model_from_file: failed to load model
llama_init_from_gpt_params: error: failed to load model 'models/xgen-4k-7b/ggml-model-f16.bin'
main: error: unable to load model
```

**Comment by tmm1** [comment]
@clyang that was with the patch to `check_vocab_size`?

I wonder instead of `return` maybe we can fill in the missing tokens with empty values

**Comment by ggerganov** [comment]
What is needed to finalize this?

**Comment by tmm1** [comment]
this seems to work and could be merged, but i'm not sure about correctness. there may be tokens missing, or a better placeholder token value (vs `b''`) for the unknown tokens.

also i haven't had a chance to look, but on the `./main` inference side i wasn't sure if it uses the embedded vocab for tokenization, and if there may be other changes required there to better conform to the differences between `tiktoken` and `sentencepiece`

File: convert.py [hunk 1]
```diff
@@ -201,6 +202,30 @@ def load(model_plus: 'ModelPlus') -> 'Params':
...
+        self.vocab_size_base: int = self.xt.vocab_size
+        self.vocab_size: int = self.xt.vocab_size
+        self.added_tokens_list = []
+
+    def all_tokens(self) -> Iterable[Tuple[bytes, float]]:
+        for index in range(0, self.vocab_size_base):
+            token = self.xt._convert_id_to_token(index)
+            yield (token, float(index))
+        for index in range(self.vocab_size_base, self.vocab_size):
+            yield (b'', float(index))
```
Comment (tmm1) [inline convert.py:223]:
```suggestion
            yield (b'<|unk|>', float(index))
```

🤷

Comment (clyang) [reply convert.py:223]: I've tried the latest commit and this is what I got.

```
$ python3 convert.py models/xgen-4k-7b-orig --outtype f16
Loading model file models/xgen-4k-7b-orig/pytorch_model-00001-of-00003.bin
Loading model file models/xgen-4k-7b-orig/pytorch_model-00001-of-00003.bin
Loading model file models/xgen-4k-7b-orig/pytorch_model-00002-of-00003.bin
Loading model file models/xgen-4k-7b-orig/pytorch_model-00003-of-00003.bin
params: n_vocab:51200 n_embd:4096 n_mult:256 n_head:32 n_layer:32
Writing vocab...
Traceback (most recent call last):
  File "/home/user/llama.cpp/convert.py", line 1255, in <module>
    main()
  File "/home/user/llama.cpp/convert.py", line 1250, in main
    OutputFile.write_all(outfile, params, output_type, model, vocab)
  File "/home/user/llama.cpp/convert.py", line 1041, in write_all
    of.write_vocab(vocab)
  File "/home/user/llama.cpp/convert.py", line 1022, in write_vocab
    self.fout.write(text)
TypeError: a bytes-like object is required, not 'str'
```

Comment (clyang) [reply convert.py:223]: I can confirm convert.py and quantize both work, thanks @smdesai !!

Btw, you still need to change EOS, BOS and NL token ID in llama.cpp to make it inference correctly.

Comment (tmm1) [reply convert.py:223]: Can you share what you changed in llama.cpp?

Comment (clyang) [reply convert.py:223]: @tmm1 Here is what I modified in llama.cpp:

```c
llama_token llama_token_bos() {
    return 50256;
}

llama_token llama_token_eos() {
    return 50256;
}

llama_token llama_token_nl() {
    return 198;
}
```

**Comment by gee842** [comment]
Hi, following this thread, I tried to convert to q4_0 or q4_1 but seem to be facing problems.
Here is my output:

```bash
ubuntu@ip-172-31-41-82:~/squeeze-llm/llama.cpp$ python3 convert.py models/xgen-7b-8k-base --outtype q4_1
Loading model file models/xgen-7b-8k-base/pytorch_model-00001-of-00003.bin
Loading model file models/xgen-7b-8k-base/pytorch_model-00001-of-00003.bin
Loading model file models/xgen-7b-8k-base/pytorch_model-00002-of-00003.bin
Loading model file models/xgen-7b-8k-base/pytorch_model-00003-of-00003.bin
params: n_vocab:51200 n_embd:4096 n_mult:256 n_head:32 n_layer:32
Traceback (most recent call last):
  File "/home/ubuntu/squeeze-llm/llama.cpp/convert.py", line 1255, in <module>
    main()
  File "/home/ubuntu/squeeze-llm/llama.cpp/convert.py", line 1248, in main
    model = convert_to_output_type(model, output_type)
  File "/home/ubuntu/squeeze-llm/llama.cpp/convert.py", line 1086, in convert_to_output_type
    return {name: tensor.astype(output_type.type_for_tensor(name, tensor))
  File "/home/ubuntu/squeeze-llm/llama.cpp/convert.py", line 1086, in <dictcomp>
    return {name: tensor.astype(output_type.type_for_tensor(name, tensor))
  File "/home/ubuntu/squeeze-llm/llama.cpp/convert.py", line 575, in astype
    self.validate_conversion_to(data_type)
  File "/home/ubuntu/squeeze-llm/llama.cpp/convert.py", line 586, in validate_conversion_to
    raise Exception(f"Can't turn an unquantized tensor into a quantized type ({data_type})")
Exception: Can't turn an unquantized tensor into a quantized type (QuantizedDataType(groupsize=32, have_addends=True, have_g_idx=False))

```

**Comment by ggerganov** [comment]
Maybe we should finalize this and merge it?
I won't be able to test this, so looking for some help in confirming it works and fixing if it's not

**Comment by alitariq4589** [comment]
Can one of the admins verify this patch?

**Comment by smdesai** [comment]
The code base has changed substantially since the initial PR. I can confirm with the code as of today, the Xgen model does convert and run f32, Q4_1 and Q4_0.

In addition, llama.cpp also required changes to work with the Xgen model. I'm not sure the best way to proceed with these changes and for now have put those inside an ifdef/ifndef XGEN

As for the best way to provide the changes, I'm not sure whether a separate PR would be better, to keep things local, here are the changes as patch files. Of course the Makefile would require a -DXGEN if compiling for the model.

[llama.cpp.patch](https://github.com/ggerganov/llama.cpp/files/12622262/llama.cpp.patch)
[convert.py.patch](https://github.com/ggerganov/llama.cpp/files/12622263/convert.py.patch)

**Comment by mofosyne** [comment]
Obsolete PR?

**Pull request closed without merging**